service/history/workflow_task_handler.go

```diff
@@ -211,6 +213,55 @@ func (handler *workflowTaskHandlerImpl) handleCommands(
 	return mutations, nil
 }
 
+func (handler *workflowTaskHandlerImpl) rejectUnprocessedUpdates(
+	ctx context.Context,
+	workflowTaskScheduledEventID int64,
+	workflowTaskHeartbeating bool,
+	wfKey definition.WorkflowKey,
+	workerIdentity string,
+) error {
+
+	// If server decided to fail WT (instead of completing), don't reject updates.
+	// New WT will be created, and it will deliver these updates again to the worker.
+	// Worker will do full history replay, and updates should be delivered again.
+	if handler.workflowTaskFailedCause != nil {
+		return nil
+	}
+
+	// If WT is a heartbeat WT, then it doesn't have to have messages.
+	if workflowTaskHeartbeating {
+		return nil
+	}
+
+	// If worker has just completed workflow with one of the WF completion command,
+	// then it might skip processing some updates. In this case, it doesn't indicate old SDK or bug.
+	// All unprocessed updates will be rejected with "workflow is closing" reason though.
+	if !handler.mutableState.IsWorkflowExecutionRunning() {
+		return nil
+	}
+
+	rejectedUpdateIDs, err := handler.updateRegistry.RejectUnprocessed(
+		ctx,
+		workflow.WithEffects(handler.effects, handler.mutableState))
+
+	if err != nil {
+		return err
+	}
+
+	if len(rejectedUpdateIDs) > 0 {
+		handler.logger.Warn(
+			"Workflow task completed w/o processing updates.",
+			tag.WorkflowNamespaceID(wfKey.NamespaceID),
+			tag.WorkflowID(wfKey.WorkflowID),
+			tag.WorkflowRunID(wfKey.RunID),
+			tag.WorkflowEventID(workflowTaskScheduledEventID),
+			tag.NewStringTag("worker-identity", workerIdentity),
+			tag.NewStringsTag("update-ids", rejectedUpdateIDs),
+		)
+	}
+	return nil
```

Comment:
synced offline. let's add a sanity check here to make sure there's no update that's in sent state.

Reply:
I added unit tests and comment here. Decided not to add extra check.